Fix same index tab stops exiting prematurely

jordanbtucker opened the pull request:
After testing, I'm convinced that destruction of `newSelection` does not need to be handled. Handling cursor change should be enough, and it doesn't have the side effect of breaking the tab stops.

Fixes #118

50Wliu commented:
It's odd how this change didn't produce a test failure.  Could you by any chance add a test that would fail without this change?

jordanbtucker commented:
Test added.

as-cii commented:
My bad! Apparently it was me to introduce that issue: I hope this hasn't caused too many headaches to you. Sorry for that :pray: 

@jordanbtucker: Nice job, by the way! I think we can merge this very soon.

Do you think we can reuse existing fixtures in this new test you have written?

> After testing, I'm convinced that destruction of newSelection does not need to be handled. 

I am quite sure you're right here, but could you elaborate a bit this thought, please? Why don't we need to handle `newSelection` destruction?

Thank you very much @jordanbtucker! :rainbow:

jordanbtucker commented:
@as-cii No headaches here.

When you `:consolidate-selections`, Atom destroys all but the *last* cursor added. So handling the destruction of `@cursor` is sufficient for that case. I'm not sure else what would destroy `newSelection`'s cursor, but I added code to handle that and destroy the `SnippetExpansion` only if not `@settingTabStop`. Can you think of any issues with that?

I also refactored the code to use existing test fixtures.

jordanbtucker commented on lib/snippet-expansion.coffee:
```diff
@@ -71,7 +71,8 @@ class SnippetExpansion
           @subscriptions.add newSelection.cursor.onDidChangePosition (event) => @cursorMoved(event)
-          @subscriptions.add newSelection.cursor.onDidDestroy => @destroy()
+          @subscriptions.add newSelection.cursor.onDidDestroy =>
+            @destroy() if not @settingTabStop
```
On new line so it's not read as:

```coffee
if not @settingTabStop
  @subscriptions.add newSelection.cursor.onDidDestroy => @destroy()
```

as-cii replied:
Really minor, but you could use `unless` here.

as-cii commented:
> I also refactored the code to use existing test fixtures.

Awesome!

> When you :consolidate-selections, Atom destroys all but the last cursor added. So handling the destruction of @cursor is sufficient for that case. I'm not sure else what would destroy newSelection's cursor, but I added code to handle that and destroy the SnippetExpansion only if not @settingTabStop. Can you think of any issues with that?

Actually I think is what we should do on every cursor destroy event. This would also be consistent with how `cursorMoved` works. What do you think about something like:

```diff
diff --git a/lib/snippet-expansion.coffee b/lib/snippet-expansion.coffee
index 8207cc4..e28ad8c 100644
--- a/lib/snippet-expansion.coffee
+++ b/lib/snippet-expansion.coffee
@@ -17,7 +17,7 @@ class SnippetExpansion
         @cursor.selection.insertText(snippet.body, autoIndent: false)
       if snippet.tabStops.length > 0
         @subscriptions.add @cursor.onDidChangePosition (event) => @cursorMoved(event)
-        @subscriptions.add @cursor.onDidDestroy => @destroy()
+        @subscriptions.add @cursor.onDidDestroy => @cursorDestroyed()
         @placeTabStopMarkers(startPosition, snippet.tabStops)
         @snippets.addExpansion(@editor, this)
         @editor.normalizeTabsInBufferRange(newRange)
@@ -29,6 +29,8 @@ class SnippetExpansion
     newTabStops = @tabStopsForBufferPosition(newBufferPosition)
     @destroy() unless _.intersection(oldTabStops, newTabStops).length
 
+  cursorDestroyed: -> @destroy() unless @settingTabStop
+
   placeTabStopMarkers: (startPosition, tabStopRanges) ->
     for ranges in tabStopRanges
       @tabStopMarkers.push ranges.map ({start, end}) =>
@@ -71,8 +73,8 @@ class SnippetExpansion
         else
           newSelection = @editor.addSelectionForBufferRange(range)
           @subscriptions.add newSelection.cursor.onDidChangePosition (event) => @cursorMoved(event)
-          @subscriptions.add newSelection.cursor.onDidDestroy =>
-            @destroy() if not @settingTabStop
+          @subscriptions.add newSelection.cursor.onDidDestroy => @cursorDestroyed()
           @selections.push newSelection
       markerSelected = true
 
```

I think we should be good to go after completing this. Thank you very much @jordanbtucker :fire:

50Wliu commented:
After looking into this a bit, looks like this will also resolve atom/language-c#65 and https://discuss.atom.io/t/special-characters-in-snippets/15445 (which is now realize doesn't have to do with special characters).

jordanbtucker commented:
Those changes all look good. I haven't had a chance to get to this today, but I'll make the changes in an hour or so. Thanks for all the input.

as-cii commented:
@jordanbtucker: thank you very much for this! Awesome job here :100: 

@50Wliu: can you confirm this fixes the issue you've mentioned? If yes, could you close them please?

Thanks everyone! :+1:

50Wliu commented:
I'll check once 0.188.0 comes out-haven't gotten around to trying to build from source yet :wink:.

izuzak commented:
I just checked #118, atom/language-c#65 and https://discuss.atom.io/t/special-characters-in-snippets/15445 and all seem to be fixed.

as-cii commented:
> I just checked #118, atom/language-c#65 and https://discuss.atom.io/t/special-characters-in-snippets/15445 and all seem to be fixed.

That's great to know! Thanks @izuzak and @50Wliu :bow:

izuzak commented:
Just published snippets v0.85.0 and bumped the version in atom/atom. Thanks for the great work here @jordanbtucker and @as-cii. :surfer: :surfer:
